fhir-bulkdata-webapp/src/main/java/com/ibm/fhir/bulkdata/provider/impl/FileProvider.java

```diff
@@ -79,43 +83,104 @@ public long getSize(String workItem) throws FHIRException {
+
+                    int chunkRead = 0;
 
-                String resourceStr = br.readLine();
-                int chunkRead = 0;
-                int maxRead = configuration.getImportNumberOfFhirResourcesPerRead(null);
-                while (resourceStr != null && chunkRead <= maxRead) {
-                    line++;
-                    chunkRead++;
-                    try {
-                        resources.add(FHIRParser.parser(Format.JSON).parse(new StringReader(resourceStr)));
-                    } catch (FHIRParserException e) {
-                        // Log and skip the invalid FHIR resource.
-                        parseFailures++;
-                        logger.log(Level.INFO, "readResources: " + "Failed to parse line " + line + " of [" + source + "].", e);
+                    String resourceStr = reader.readLine();
+                    boolean continueRead = true;
+                    while (continueRead && resourceStr != null) {
+                        chunkRead++;
+                        try (StringReader stringReader = new StringReader(resourceStr)){
+                            resources.add(
+                                        FHIRParser.parser(Format.JSON)
+                                                    .parse(stringReader));
+                        } catch (FHIRParserException e) {
+                            // Log and skip the invalid FHIR resource.
+                            parseFailures++;
+                            logger.log(Level.INFO, "readResources: Failed to parse line " + (numOfLinesToSkip + chunkRead) + " of [" + source + "].", e);
+                        }
+
+                        // With small resources the 50 at a time gets really expensive.
+                        // so we have a READ_BLOCK_OPT to get at least a decent segment to
+                        // insert into the db.
+                        if (chunkRead < maxRead || reader.getLength() < READ_BLOCK_OPT) {
+                            resourceStr = reader.readLine();
+                        } else {
+                            resourceStr = null;
+                            continueRead = false;
+                        }
                     }
-                    resourceStr = br.readLine();
-                }
-            } finally {
-                if (br != null) {
-                    br.close();
+                    // The Channel enables us to skip the future seek.
+                    this.transientUserData.setCurrentBytes(this.transientUserData.getCurrentBytes() + reader.getLength());
                 }
             }
         } catch (Exception e) {
             throw new FHIRException("Unable to read from Local File", e);
         }
     }
 
+    /**
+     * This specialized class enables a BufferedReader.readLine like implementation
+     * without the buffering as there is a local file with direct read and seek.
+     */
+    private static class CountInputStreamReader extends InputStreamReader {
```

Comment:
What about multi-byte characters? I think this needs to be based of a Reader, not an InputStream.